biggraphite/accessor.py

```diff
@@ -0,0 +1,456 @@
+#!/usr/bin/env python
+# Copyright 2016 Criteo
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Abstracts querying Cassandra to manipulate timeseries."""
+from __future__ import print_function
+
+import json
+
+import cassandra
+from cassandra import cluster as c_cluster
+from cassandra import concurrent as c_concurrent
+from cassandra import encoder as c_encoder
+import statistics
+
+
+class Error(Exception):
+    """Base class for all exceptions from this module."""
+
+
+class CassandraError(Error):
+    """Fatal errors accessing Cassandra."""
+
+
+class RetryableCassandraError(Error):
+    """Errors accessing Cassandra that could succeed if retried."""
+
+
+class InvalidArgumentError(Error):
+    """Callee did not follow requirements on the arguments."""
+
+
+class InvalidGlobError(InvalidArgumentError):
+    """The provided glob is invalid."""
+
+
+class TooManyMetrics(InvalidArgumentError):
+    """A name glob yielded more than Accessor.MAX_METRIC_PER_GLOB metrics."""
+
+
+_LAST_COMPONENT = '__END__'
+
+_COMPONENTS_MAX_LEN = 64
+
+_SETUP_CQL_METRICS_COMPONENTS = "\n".join(
+    "  component_%d text," % n for n in range(_COMPONENTS_MAX_LEN)
+)
+
+_SETUP_CQL_METRICS_INDEXES = [
+    "CREATE CUSTOM INDEX IF NOT EXISTS ON metrics (component_%d)"
+    "  USING 'org.apache.cassandra.index.sasi.SASIIndex'"
+    "  WITH OPTIONS = {"
+    "    'analyzer_class': 'org.apache.cassandra.index.sasi.analyzer.NonTokenizingAnalyzer',"
+    "    'case_sensitive': 'true'"
+    "  };" % n for n in range(_COMPONENTS_MAX_LEN)
+]
+
+_SETUP_CQL = [
+    "CREATE TABLE IF NOT EXISTS raw ("
+    "  metric text,"
+    "  time_start_ms bigint,"
+    "  time_offset_ms int,"
+    "  value double,"
+    # TODO: Add aggregation_step
+    "  PRIMARY KEY ((metric, time_start_ms), time_offset_ms)"
+    ")"
+    "  WITH CLUSTERING ORDER BY (time_offset_ms DESC)"
+    "  AND compaction={"
+    "    'timestamp_resolution': 'MICROSECONDS',"
+    "    'max_sstable_age_days': '1',"
+    "    'base_time_seconds': '900',"
+    "    'class': 'DateTieredCompactionStrategy'"
+    "  };",
+    "CREATE TABLE IF NOT EXISTS metrics ("
```

Comment:
/metrics/metadata/ ?
Can we also add something to distinguish directory from files ?